src/main/java/io/r2dbc/h2/H2ConnectionMetadata.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.h2;
+
+import io.r2dbc.spi.ConnectionMetadata;
+import org.h2.engine.Constants;
+
+/**
+ * Connection metadata for a connection connected to a H2 database.
+ */
+public enum H2ConnectionMetadata implements ConnectionMetadata {
+
+    INSTANCE;
+
+    @Override
+    public String getDatabaseProductName() {
+        return "H2";
+    }
+
+    @Override
+    public String getDatabaseVersion() {
+        return Constants.getFullVersion();
```

Comment:
The Javadoc states
> Retrieves the version number of this database product.

But this constant actually returns a version of the client driver, the server can have a different version. Perhaps the documentation (in spi project) should be clear for end users about the behavior of this method in such situation. Maybe you need two different methods.

Reply:
Do you mean, if it was a client-server H2 setup, the client could be at one version and the server another? If so, how would you get the server's version through the client? Is there an API to look that up?

Reply:
Yes, they can. To determine the version of the server you need to execute `CALL H2VERSION()` and possibly cache its result, if you wish.

Reply:
Note that this function returns `Constant.getVersion()` on the server side (a version without a build date), so you may want to use this method instead of `getFullVersion()` for a client version too.

Reply:
BTW, JDBC has separate versions of the client driver (in `Driver` class) and of the database (in `DatabaseMetaData` interface), but not all drivers return real values. But in case of R2DBC the version of H2's client driver looks like something unnecessary, because its API isn't used directly by the application. However, if an old driver is used some features may be unavailable. For example, old drivers don't know the `INTERVAL` data type, so the server reports such values as `VARCHAR` values to them instead.

Reply:
`CALL H2VERSION()` seems quite appealing. I think I'm going to change my PR to use exactly that.

Reply:
@mp911de Invoke via SQL, or call directly into `Constant.getVersion()`?

Reply:
You need SQL to get server's version.